src/common/libhostlist/hostlist.c

```diff
@@ -0,0 +1,1140 @@
+/************************************************************\
+ * Copyright 2020 Lawrence Livermore National Security, LLC
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif /* !HAVE_CONFIG_H */
+
+#include <string.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+#include <assert.h>
+#include <errno.h>
+#include <ctype.h>
+#include <sys/param.h>
+#include <unistd.h>
+
+#include "hostlist.h"
+#include "hostrange.h"
+#include "util.h"
+
+/* number of elements to allocate when extending the hostlist array */
+#define HOSTLIST_CHUNK    16
+
+/* max host range: anything larger will be assumed to be an error */
+#define MAX_RANGE    16384    /* 16K Hosts */
+
+/* max host suffix value */
+#define MAX_HOST_SUFFIX 1<<25
+
+/* max number of ranges that will be processed between brackets */
+#define MAX_RANGES    10240    /* 10K Ranges */
+
+/* size of internal hostname buffer (+ some slop), hostnames will probably
+ * be truncated if longer than MAXHOSTNAMELEN */
+#ifndef MAXHOSTNAMELEN
+#define MAXHOSTNAMELEN    64
+#endif
+
+/* max size of internal hostrange buffer */
+#define MAXHOSTRANGELEN 1024
+
+/* Helper structure for hostlist iteration
+ */
+struct current {
+    char *host;
+    int index;
+    int depth;
+};
+
+/* Hostlist - a dynamic array of hostrange objects
+ */
+struct hostlist {
+    int size;               /* current number of elements available in array */
+    int nranges;            /* current number of ranges stored in array */
+    int nhosts;             /* current number of hosts stored in hostlist */
+    struct hostrange **hr;  /* pointer to hostrange array */
+
+    struct current current; /* iterator cursor */
+};
+
+/* _range struct helper for parsing hostlist strings
+ */
+struct _range {
+    unsigned long lo, hi;
+    int width;
+};
+
+
+/*
+ * Helper function for host list string parsing routines
+ * Returns a pointer to the next token; additionally advance *str
+ * to the next separator.
+ *
+ * next_tok was taken directly from pdsh courtesy of Jim Garlick.
+ * (with modifications to support bracketed hostlists, i.e.:
+ *  xxx[xx,xx,xx] is a single token)
+ *
+ * next_tok now handles multiple brackets within the same token,
+ * e.g.  node[01-30]-[1-2,6].
+ */
+static char * next_tok(char *sep, char **str)
```

Comment:
flux code style nit: space after function name

Reply:
Yeah, I did try to run `clang-format` to automatically update some of these coding style issues, but gave up after spending about 40 minutes on it.

Since we already used to have the previous version of the hostlist code in flux-core, I didn't want to spend a few hours updating all style and other minor issues in the code, but only fixed what I came across while developing the unit tests.

In the interest of saving time, we should perhaps avoid doing a full code review at this time.